Frontend structure, i18n, and env improvements

FarukCaluk opened the pull request:
This pull request introduces several key improvements to the codebase, focusing on internationalization support, API integration for authentication and user management, and code organization. The most significant changes are the addition of i18n for multi-language support, new API modules for user and auth operations, and refactoring of file and import paths for consistency.

**Internationalization (i18n) Support:**

* Added `i18next` and `react-i18next` dependencies and created the `src/i18n/i18n.ts` configuration file, enabling English and Bosnian language support throughout the app. UI components like `Header` now use translation keys and allow runtime language switching. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R16-L22) [[2]](diffhunk://#diff-81061d307e416e4e3ed196be2a730aa92209f2c4ad9624f0121c58abe913be30R1-R82) [[3]](diffhunk://#diff-f5f4277f54f0ec5d48c5a0d4102a1142c507e02a91ce84525e6e8fda4fee6231R3-L6) [[4]](diffhunk://#diff-f5f4277f54f0ec5d48c5a0d4102a1142c507e02a91ce84525e6e8fda4fee6231L18-R45) [[5]](diffhunk://#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405L3-R6)

**API Integration:**

* Introduced new API modules for authentication (`src/api/auth.ts`) and admin user management (`src/api/admin/users.ts`), providing functions for login, registration, password reset, user listing, creation, editing, deletion, and blocking. [[1]](diffhunk://#diff-737133f9502a5031455242f2da0022178f68c346dcd5f1c5f47547ffb5a06c0aR1-R36) [[2]](diffhunk://#diff-4937285c2d1210b2f5924375f6eab1a4a2f084debdc5034adf1a15595632ae9cR1-R35)

**Code Organization and Refactoring:**

* Standardized file and import naming conventions (e.g., changed imports from `Layout` to `layout`, from `Users` to `users`, etc.) and moved hooks from `api/hooks` to `hooks` for improved structure. [[1]](diffhunk://#diff-1deef93890e359e3ea9cccd8b9fb992a1008c1f21b9eee105a1a364afae4f170L5-R9) [[2]](diffhunk://#diff-3d2518b4a3eba81f0f0faab4a2484491b0d509acf014cf7f2886fe248ad8cb85L2-R3) [[3]](diffhunk://#diff-52d7b60afcd61a0be32e8122adc0c6f3564188e502e1059459498fc6ab84edd4L4-R4) [[4]](diffhunk://#diff-4655d8451c3ba2274a2424fbd04d9bab747b2ddcdb17f671b452381e5e272864L13-R13)

**Environment Configuration:**

* Added `API_URL` to `.env`, `.env.development`, and `.env.example` files to centralize API endpoint configuration. [[1]](diffhunk://#diff-e9cbb0224c4a3d23a6019ba557e0cd568c1ad5e1582ff1e335fb7d99b7a1055dR1) [[2]](diffhunk://#diff-a3046da0d15a27e89f2afe639b25748a7ad4d9290af3e7b1b6c1a5533c8f0a8cR1)

**Dependency Updates:**

* Upgraded `typescript` to version 5.9.3 and moved its declaration to `devDependencies` in `package.json`.

Let me know if you want details on any specific change or how to use the new i18n or API modules!

ttare commented on src/api/admin/users.ts:
```diff
@@ -0,0 +1,36 @@
+};
+
+export const editUser = async (id: string, user: any) => {
+  const res = await fetch(`${process.env.API_URL}/admin/users/${id}`, {
```
Does it make sense to write this every time you need to call api? Do you have idea how to simplify the call with few helper functions?

FarukCaluk replied:
Yes, I agree — repeating the fetch logic for every API call is not the best approach. I plan to simplify this by creating a helper (e.g., apiClient based on axios) where the baseURL and default headers will already be defined. That way in users.ts and other files I’ll just use apiClient.get(...), apiClient.post(...), etc. This will make the code cleaner and easier to maintain.

ttare commented on src/i18n/i18n.ts:
```diff
@@ -0,0 +1,82 @@
+i18n.use(initReactI18next).init({
+  resources: {
+    en: {
+      translation: {
```
This file will contain translations for 10-15 pages. We should split it. Also think about better way to organize translations json by specific feature. Ask chat gpt about potential solutions.

FarukCaluk replied:
That makes sense — having all translations in a single file will get messy once we have more pages. A better approach would be to split the translations by feature or module, for example:

/src/i18n/locales/en/users.json  
/src/i18n/locales/en/simcards.json  
/src/i18n/locales/en/auth.json  


Then in i18n.ts we can dynamically load them and merge resources. This way, each page or feature owns its translations, making it easier to maintain and extend in the future.

I’ll research some best practices (namespacing, lazy loading JSON files) and propose a structure we can stick to.

ttare replied:
For me it is fine to have all translation in one file. You don't have to use namespaces but its good to know that they exists. Also take care for naming translations. "TABLE.ORDER_NAME" is not good as "ORDER.TABLE.NAME". Pattern could be something like "PAGE.COMPONENT.PROP" or "PAGE.PROP". In that case we have to group JSON file in a way to have few lines break for each new page

ttare commented:
@FarukCaluk Great work and description in the PR details. Please apply following

1. once PR is created set me as reviewer so I get notification that PR is created 
2. regarding api layer now you have 2 instances.
2.1 Axios and Fetch. Please read about these two and main differences and chose one.
3. decide tailwind vs bootstrap
4. Its important to think about user roles. Can you write simple documentation where you will write roles and describe what each role will have access to, what will be display and so on. For example access to users will be only allowed to our team an nobody else. So we have to describe role access and regarding that organize project structure.

FarukCaluk commented:
Summary of changes:

Restructured pages by user roles (admin, support, agency, user, anonymous)
Created router components for each role
Updated App.tsx to use role-based routers
Updated sidebar to show only allowed links per role
Removed PrivateRoute component and its usages

ttare commented on src/App/App.tsx:
```diff
@@ -1,57 +1,24 @@
-        {/* Default: redirect to /login */}
-        <Route path="*" element={<LoginPage />} />
-      </Routes>
+      {!isLoggedIn && <AnonymousRouter />}
```
This is much better to see :D On more thing to improve readability is following.

if (is not logged in ) { 
   return <AnonymousRouter />
}

switch (role) {
....

}

ttare commented on src/api/admin/users.ts:
```diff
@@ -0,0 +1,26 @@
+  return res.data;
+};
+
+export const createUser = async (user: any) => {
```
do not use "any". I know that you don't know what all props this will required but you can create type called CreateUserPayload for example that is empty and we will extend it.

Also each function need to return appropriate type

ttare commented on src/api/auth.ts:
```diff
@@ -0,0 +1,35 @@
+export const login = async (credentials: any) => {
+  const res = await fetch(`${process.env.API_URL}/auth/login`, {
```
Didn't we agree that you will use axios ? So no mixing

ttare commented on src/api/models/collection-modul.ts:
```diff
@@ -0,0 +1,7 @@
+import api from '../../utils/api-client';
+import { Collection } from '../../types/collection.types';
+
+export const getCollections = async (): Promise<Collection[]> => {
```
This is 👌

ttare commented on src/components/Layout/Sidebar.tsx:
```diff
@@ -16,22 +16,70 @@ const Sidebar: React.FC<{ open: boolean }> = ({ open }) => {
-              to="/simcards"
-              className={`nav-link${location.pathname === '/simcards' ? ' active' : ''}`}
+              to="/dashboard"
+              className={`nav-link${location.pathname === '/dashboard' ? ' active' : ''}`}
```
you can take a look for the helper library that will make working with classes (in this way) looks better.

ttare commented on src/components/Layout/Sidebar.tsx:
```diff
@@ -16,22 +16,70 @@ const Sidebar: React.FC<{ open: boolean }> = ({ open }) => {
             </Link>
           </Nav.Item>
+          {/* SIM Cards: admin, support, agency */}
+          {(role === 'admin' || role === 'support' || role === 'agency') && (
```
1. Use Enums for Roles. In case we change name or add new role it will be easier to maintain.

2. There are two approachs for NavItems.
2.1 If you decide to manually add each sidebar item (I don't recommend that) then this file will contain hundreds of lines. So to decrease it you have to have your custom nav item <NavItem title={t('Sidebar...')} path="/simcards" active={true / false}  />

2.2 Better solution was to extend it with role specific array and dynamicaly render sidebar items 

const items: Record<Role, SidebarItem> = {
'admin': [{ name:  'sidebar.users', path: '/admin/users', icon 'users }]

}


}

ttare commented on src/i18n/i18n.ts:
```diff
@@ -0,0 +1,51 @@
+import bsPackages from './locales/bs/packages.json';
+import bsCollections from './locales/bs/collections.json';
+import bsPlans from './locales/bs/plans.json';
+import bsDashboard from './locales/bs/dashboard.json';
```
To avoid all these imports take a look for the another solution. Check docs for the library

ttare commented on src/i18n/i18n.ts:
```diff
@@ -0,0 +1,51 @@
+import i18n from 'i18next';
+import { initReactI18next } from 'react-i18next';
+
```
in this file define supported languages array so we have it in one place.

ttare commented on src/pages/anonymous/login/login-page.tsx:
```diff
@@ -16,22 +16,23 @@ export const LoginPage: React.FC = () => {
       const response = await fetch('http://localhost:3000/api/auth/login', {
```
You have this function somewhere else so its repeated

ttare commented on src/pages/anonymous/login/login-page.tsx:
```diff
@@ -16,22 +16,23 @@ export const LoginPage: React.FC = () => {
-      // Redirect to /simcards
-      navigate('/simcards');
+      navigate('/dashboard');
+      window.location.reload();
```
This is the something you will must fix. But leave it as last comment becasue maybe wou will need help for this. Reload aplication state without using window.reload (full page refresh)

FarukCaluk commented:
Summary of changes:
- Restructured pages by role and added folder entry points (index.tsx) for admin/support/agency/user/anonymous and feature subfolders.
- Replaced string role checks with a typed Role enum.
- Switched auth to axios client (src/api/auth.ts) and updated login flow to notify App (no window.reload).
- Added typed API payloads/returns (src/types/user.ts) and updated src/api/admin/users.ts (removed `any`).
- Updated Sidebar/Header to use Role enum and kept i18n to English only.
- Removed PrivateRoute and adjusted routing to role-based routers (AdminRouter, SupportRouter, AgencyRouter, UserRouter, AnonymousRouter).

Notes / requests:
- Please review naming/index exports and role-based routing structure.
- I left CreateUserPayload minimal — we can extend types as needed.
- I removed Bosnian translations per request; we can generate translations later.

Ready to apply additional changes (AuthContext, sidebar data-driven refactor, or further typing) if you request them. Please review and advise.

FarukCaluk commented:
Structure improvements: Tailwind UI, editable profile
Summary
Migrated layout to Tailwind and refreshed UI.
Sidebar now collapses to an icon rail (navigation stays clickable); logout moved next to sidebar.
Added Edit Profile flow so changes persist.
